src/main/java/org/commcare/suite/model/MenuLoader.java

```diff
@@ -145,11 +146,28 @@ private boolean menuIsRelevant(SessionWrapperInterface sessionWrapper, Menu m) t
         return true;
     }
 
-    public boolean menuAssertionsPass(SessionWrapperInterface sessionWrapper, Menu m) {
-        Text text =  m.getAssertions().getAssertionFailure(sessionWrapper.getEvaluationContext());
-        if (text != null) {
-            loadException = new Exception(text.evaluate());
-            return false;
+    public boolean menuAssertionsPass(SessionWrapperInterface sessionWrapper, Menu m) throws XPathSyntaxException{
+        Vector<String> assertionXpathStrings = m.getAssertions().getAssertionsXPaths();
+        if (!assertionXpathStrings.isEmpty()) {
+            XPathExpression assertionXpath = XPathParseTool.parseXPath(assertionXpathStrings.get(0));
+
+            Set<String> instancesNeededByAssertionCondition =
+                    (new InstanceNameAccumulatingAnalyzer()).accumulate(assertionXpath);
+            EvaluationContext ec = sessionWrapper.getRestrictedEvaluationContext(m.getId(),
+                    instancesNeededByAssertionCondition);
+
+            EvaluationContext traceableContext = new EvaluationContext(ec, ec.getOriginalContext());
+            if (traceReporter != null) {
+                traceableContext.setDebugModeOn(traceReporter);
+            }
+
+            InstrumentationUtils.printAndClearTraces(traceReporter, "menu load expand");
+            Text text =  m.getAssertions().getAssertionFailure(ec);
```

Comment:
is it possible to not repeat this code for assertions and relevancy xpaths by moving it to different method.

Reply:
I was working on breaking out the debug code into a separate function, but then I realized you might be referring to that and the instance accumulation code as well. Do you mean both?